### PR TITLE
Add import.meta.main

### LINKED ISSUE #1834: Add import.meta.main
Currently we can use `import.meta.url == window.location.toString()` to test if a given script is the main script... Similar to `require.main == module` in Node.

But `import.meta.url == window.location.toString()` is quite a bit to type, and `import.meta` is allowed to be extended... https://github.com/tc39/proposal-import-meta

I'd like to add `import.meta.main`, a boolean, as a short hand to decide if a module is the main.

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #1834

~~I tried to move `ImportMeta` declaration to separate, however TS was crushing then, I'll investigate that tomorrow.~~

I had also test in C++, but it felt duplicated with TS test.

### CODE REVIEW

**Comment by kitsonk** [comment]
As this stands at the moment, it won't flow through to the runtime type library.  (If you do `deno --types` you will see it is missing).

The interface is already a global interface (`interface ImportMeta {}`) as it has been reserved in the language for a long time.  We need to extend this interface in the global scope, which is what happens when it is located in `lib.web_assembly.d.ts` file as that is simply inlined in the runtime type library, but that isn't the "right" way to do it for extending this in the long term.  I think it is better that we put this directly in the `globals.ts`, the only problem is that we need to change `ts_library_builder` in the `mergeGlobal` function to copy over any interfaces there into the type library.

That is a bit complicated, so if you wanted to, just move it back to the web assembly library, and then I can do a follow up PR making the change to `ts_library_builder`.  Otherwise you can try to tackle it.  It should be added somewhere [around here](https://github.com/denoland/deno/blob/e109941bd256b276ea3a68717bf519a5b3a08ff1/tools/ts_library_builder/build_library.ts#L273) and needs to be done is iterate over any interfaces, ensuring it weeds out the special `global` interface and adds those to the target.

File: libdeno/binding.cc [hunk 1]
```diff
@@ -520,7 +521,19 @@ void HostInitializeImportMetaObjectCallback(v8::Local<v8::Context> context,
 
   const char* url = info->name.c_str();
 
+  // check if `module` is "main" module
+  // ie. `import.meta.url` === `window.location.toString()`
+  auto location_obj = context->Global()->Get(v8_str("location"));
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply libdeno/binding.cc]: I agree, in cpp tests `window.location` is undefined because Deno doesn't set it, I added the test but it's green on my machine 😅

**Comment by bartlomieju** [comment]
@kitsonk actually I get:
```
$ deno --types | grep -i importmeta -A 3 -B 3
```
```ts
// @url js/lib.import_meta.d.ts

interface ImportMeta {
  url: string;
  main: boolean;
}
```

**Comment by ry** [comment]
Maybe we can add an argument to `deno_mod_new()` to tell if it's main...

**Comment by bartlomieju** [comment]
@ry I can place it in `ModuleInfo` - `ModuleInfo.main`?

**Comment by ry** [comment]
@bartlomieju yes that makes sense to me

**Comment by bartlomieju** [comment]
@ry not super happy because there's a lot of bool flags passing but it got job done

File: BUILD.gn [hunk 2]
```diff
@@ -199,7 +200,8 @@ run_node("deno_runtime_declaration") {
     "--basePath",
     rebase_path(".", root_build_dir),
     "--inline",
-    rebase_path("js/lib.web_assembly.d.ts", root_build_dir),
+    rebase_path("js/lib.import_meta.d.ts", root_build_dir) + "," +
+        rebase_path("js/lib.web_assembly.d.ts", root_build_dir),
```
Comment (ry) [inline BUILD.gn]: Don't do this in this commit. Let's do it in a separate PR. Just leave it in `lib.web_assembly.d.ts` for now.

File: libdeno/modules_test.cc [hunk 3]
```diff
@@ -266,3 +266,31 @@ TEST(ModulesTest, ImportMetaUrl) {
   EXPECT_EQ(nullptr, deno_last_exception(d));
   EXPECT_EQ(1, exec_count);
 }
+
+TEST(ModulesTest, ImportMetaMain) {
+  Deno* d = deno_new(deno_config{0, empty, empty, recv_cb});
+
+  const char* a_src = "if (!import.meta.main) throw 'hmm'";
+  static deno_mod a = deno_mod_new(d, true, "a.js", a_src);
```
Comment (ry) [inline libdeno/modules_test.cc]: rename `a_src` to `throw_not_main_src` 
rename `a` to `throw_not_main`

rename `b_src` to `throw_main_src` 
rename `b` to `throw_main`

File: libdeno/modules_test.cc [hunk 4]
```diff
@@ -266,3 +266,31 @@ TEST(ModulesTest, ImportMetaUrl) {
...
+  EXPECT_NE(a, 0);
+  EXPECT_EQ(nullptr, deno_last_exception(d));
+
+  deno_mod_instantiate(d, d, a, nullptr);
+  EXPECT_EQ(nullptr, deno_last_exception(d));
+
+  deno_mod_evaluate(d, d, a);
+  EXPECT_EQ(nullptr, deno_last_exception(d));
+
+  const char* b_src = "if (import.meta.main) throw 'hmm'";
```
Comment (ry) [inline libdeno/modules_test.cc]: s/hmm/err/

File: tests/import_meta.ts.out [hunk 5]
```diff
@@ -1,2 +1,2 @@
-import_meta2 [WILDCARD]import_meta2.ts
-import_meta [WILDCARD]import_meta.ts
+import_meta2 [WILDCARD]import_meta2.ts false
+import_meta [WILDCARD]import_meta.ts true
```
Comment (ry) [inline tests/import_meta.ts.out]: 👍

**Comment by bartlomieju** [comment]
@ry changes applied

File: js/lib.web_assembly.d.ts [hunk 6]
```diff
@@ -162,7 +162,7 @@ declare namespace WebAssembly {
   }
 }
 
-// TODO Move ImportMeta intos its own lib.import_meta.d.ts file?
```
Comment (ry) [inline js/lib.web_assembly.d.ts]: Leave this TODO please

Comment (bartlomieju) [reply js/lib.web_assembly.d.ts]: sorry, my mistake

File: libdeno/modules_test.cc [hunk 7]
```diff
@@ -266,3 +266,32 @@ TEST(ModulesTest, ImportMetaUrl) {
...
+  EXPECT_NE(throw_main, 0);
+  EXPECT_EQ(nullptr, deno_last_exception(d));
+
+  deno_mod_instantiate(d, d, throw_main, nullptr);
+  EXPECT_EQ(nullptr, deno_last_exception(d));
+
+  deno_mod_evaluate(d, d, throw_main);
+  EXPECT_EQ(nullptr, deno_last_exception(d));
+
+  deno_delete(d);
```
Comment (ry) [inline libdeno/modules_test.cc]: Nice test!

**Comment by bartlomieju** [comment]
Thanks @ry, somehow missed that file...

**Review by Fool256 — APPROVED** [review]
